src/tableau_server/tableau_server_objects.py

```diff
@@ -0,0 +1,372 @@
+        self.__to_int('contents_counts_view_count')
+        self.__to_int('contents_counts_workbook_count')
+        self.__to_int('favorites_total')
+        self.__to_int('finish_code')
+        self.__to_int('port')
+        self.__to_int('priority')
+        self.__to_int('server_port')
+        self.__to_int('sheet_count')
+        self.__to_int('size')
+        self.__to_int('usage_total_view_count')
+        self.__to_int('user_count')
+
+    def __hash__(self):
+        return hash(str(astuple(self)))
+
+    def to_dict(self):
+        return asdict(self)
+
+
+@dataclass
+class User(ServerObject):
+    """ A Tableau User object """
+    name: str
+    full_name: str = None
+    domain_name: str = None
+    auth_setting: str = None
+    email: str = None
+    email_domain: str = None
+    external_auth_user_id: str = None
+    site_role: str = None
+    locale: str = None
+    language: str = None
+    last_login: datetime = None
+
+    def __post_init__(self):
+        if not self.email_domain and self.email:
+            self.email_domain = self.email.lower().split('@')[-1]
+        super().__post_init__()
+
+    def __hash__(self):
+        return hash(str(astuple(self)))
+
+
+@dataclass
+class Project(ServerObject):
+    """ A Tableau Project object """
+    name: str
+    owner_id: str = None
+    parent_project_id: str = None
+    description: str = None
+    controlling_permissions_project_id: str = None
+    created_at: datetime = None
+    updated_at: datetime = None
+    content_permissions: str = None
+    top_level_project: bool = False
+    writeable: bool = False
+    contents_counts_project_count: int = None
+    contents_counts_view_count: int = None
+    contents_counts_datasource_count: int = None
+    contents_counts_workbook_count: int = None
+
+    def __hash__(self):
+        return hash(str(astuple(self)))
+
+
+@dataclass
+class Group(ServerObject):
+    """ A Tableau Group object """
+    name: str
+    domain_name: str = None
+    import_domain_name: str = None
+    import_site_role: str = None
+    import_grant_license_mode: str = None
+    minimum_site_role: str = None
+    user_count: int = None
+
+    def __hash__(self):
+        return hash(str(astuple(self)))
+
+
+@dataclass
+class Datasource(ServerObject):
+    """ A Tableau Datasource object """
+    name: str
+    created_at: datetime = None
+    updated_at: datetime = None
+    project_id: str = None
+    project_name: str = None
+    owner_id: str = None
+    tags: list = None
+    description: str = None
+    content_url: str = None
+    type: str = None
+    encrypt_extracts: bool = False
+    has_alert: bool = False
+    has_extracts: bool = False
+    is_certified: bool = False
+    is_published: bool = True
+    use_remote_query_agent: bool = False
+    database_name: str = None
+    server_name: str = None
+    favorites_total: int = None
+    connected_workbooks_count: int = None
+
+    def publish_xml(self, conn_creds=None):
```

Comment:
This looks to be using the username and PW authentication.  I will be using API token authentication. Is that built in here or we'd have to add it

Reply:
Hmm I have not tried to do that, we could perhaps pair on this at some point to test it out and make it work 😄